### PR TITLE
Rule skip

### PR DESCRIPTION
**Pull request by SFJohnson24**
[Rule_underscoresidvar.json](https://github.com/user-attachments/files/22009170/Rule_underscoresidvar.json)
this PR updates CG0371 with rule logic above.  This allows IDVAR and IDVARVAL to be matched for child->parent merges
to test:
run new rule logic with test cases

This pull request introduces significant improvements to the dataset preprocessing logic, particularly for handling supplemental datasets and merging records based on IDVAR/IDVARVAL keys. The changes enhance the robustness and flexibility of merging child and parent datasets, ensuring accurate record matching even in complex scenarios. Additionally, new comprehensive tests have been added to validate these behaviors.

### Enhancements to Dataset Preprocessing and Merging Logic

* Refactored the `_child_merge_datasets` method in `dataset_preprocessor.py` to classify datasets as supplemental or RDOMAIN and apply specialized merging logic for cases involving IDVAR/IDVARVAL keys. This includes new helper methods for classification and merging, improving handling of supplemental datasets and edge cases. [[1]](diffhunk://#diff-4b1636a769d953b60785e877646b5b1c1cc6a1aade0a596c4979b9f96feb18faL190-R206) [[2]](diffhunk://#diff-4b1636a769d953b60785e877646b5b1c1cc6a1aade0a596c4979b9f96feb18faR226-R459)
* Added support for type-safe comparisons between IDVARVAL and parent dataset columns, including fallback mechanisms for string conversion and error handling, ensuring robust matching across different data types.

### API and Interface Improvements

* Updated function signatures to use explicit typing for `match_keys` and added the `datasets` parameter, improving code clarity and maintainability. [[1]](diffhunk://#diff-4b1636a769d953b60785e877646b5b1c1cc6a1aade0a596c4979b9f96feb18faR138) [[2]](diffhunk://#diff-4b1636a769d953b60785e877646b5b1c1cc6a1aade0a596c4979b9f96feb18faL190-R206)
* Removed unnecessary logging of preprocessed datasets to reduce noise in logs.

### Testing and Validation

* Added multiple new unit tests to `test_dataset_preprocessor.py` covering supplemental dataset merging, integer IDVAR matching, cases with no matches, and combined standard and IDVAR matching. These tests ensure correctness and cover a wide range of scenarios.

### CODE REVIEW

File: cdisc_rules_engine/utilities/dataset_preprocessor.py [hunk 1]
```diff
@@ -187,30 +188,274 @@ def _child_merge_datasets(
...
-            "--",
-            left_dataset_domain_name,
+        is_supplemental, rdomain_dataset = self._classify_dataset(
+            left_dataset, self._dataset_metadata
         )
-        right_dataset_match_keys = replace_pattern_in_list_of_strings(
-            get_sided_match_keys(match_keys=match_keys, side="right"),
-            "--",
+        has_idvar_keys = "IDVAR" in match_keys or "IDVARVAL" in match_keys
+        if is_supplemental or rdomain_dataset and has_idvar_keys:
```
Comment (RamilCDISC) [inline cdisc_rules_engine/utilities/dataset_preprocessor.py]: Could you please explain if here the intent is is_supplemental or (rdomian_dataset and has_idvar_keys) or (is_supplemental or rdomian_dataset) and has_idvar_keys?

Comment (SFJohnson24) [reply cdisc_rules_engine/utilities/dataset_preprocessor.py]: forgot the parens.  goal was for it to be searching for dataset that is supplemental/rdomain containing (RELREC/CO) and having idvar/idvarval in the match keys.   (is_supplemental or rdomian_dataset) and has_idvar_keys

**Review by RamilCDISC — APPROVED** [review]
The PR updates the dataset preprocessor to account for the cases related to IDVAR/IDVARVAL as described in the description. The PR was validated by:

1. Reviewing the updated code for any unwanted code or comments. 
2. Reviewing the updated logic to validate consistency and that it fulfills the requirements.
3. Verifying the relevance tests are updated with proper edge cases.
4. Validating passing of all unit tests.
5. Validating passing of end2end testing to ensure editor is not broken by the update.
6. Validating the passing of test_suite to verify proper integration of updated code.
7. Validating passing of all regression testing.
8. Manually running a validation using the rule attached in the PR and CG0371 positive and negative datasets.
